### PR TITLE
validating the date format, add test case, since NumberFormatException extends IllegalArgumentException, it is only necessary to write IllegalArgumentException.

### PR DESCRIPTION
**Pull request by Carpe-Wang**
validating the date format, add test case and since NumberFormatException extends IllegalArgumentException, it is only necessary to write IllegalArgumentException.

#### Purpose

* Throw exception if  `public GsonBuilder setDateFormat(String pattern) ` input pattern is invalid. [[Relate issue](https://github.com/google/gson/issues/2472)](https://github.com/google/gson/issues/2472)
* NumberFormatException extends IllegalArgumentException, it is only necessary to write IllegalArgumentException in `public class ISO8601Utils`

### Checklist

- [x] New code follows the [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html)
- [x] If necessary, new public API validates arguments, for example rejects `null`
- [x] New public API has Javadoc
    - [x] Javadoc uses `@since $next-version$`  
      (`$next-version$` is a special placeholder which is automatically replaced during release)
- [x] If necessary, new unit tests have been added  
  - [x] Assertions in unit tests use [Truth](https://truth.dev/), see existing tests
  - [x] No JUnit 3 features are used (such as extending class `TestCase`)
  - [x] If this pull request fixes a bug, a new test was added for a situation which failed previously and is now fixed
- [x] `mvn clean verify javadoc:jar` passes without errors

### CODE REVIEW

**Review by eamonnmcmanus — CHANGES_REQUESTED** [review]
This looks great, thanks! Could you add a test, for example in `DefaultTypeAdaptersTest`, that fails with the current code and passes with this change? You could use `assertThrows`: with the current code the test will fail because no exception will be thrown.

**Comment by Carpe-Wang** [comment]
> This looks great, thanks! Could you add a test, for example in `DefaultTypeAdaptersTest`, that fails with the current code and passes with this change? You could use `assertThrows`: with the current code the test will fail because no exception will be thrown.

Yes, I add them.

File: gson/src/test/java/com/google/gson/functional/DefaultTypeAdaptersTest.java
Comment (eamonnmcmanus) [inline gson/src/test/java/com/google/gson/functional/DefaultTypeAdaptersTest.java:722]: Just one tiny thing: could you add a blank line before each of these new test methods?

Comment (Carpe-Wang) [reply gson/src/test/java/com/google/gson/functional/DefaultTypeAdaptersTest.java:722]: OK, I am sorry about that.

**Comment by Carpe-Wang** [comment]
> This looks great, thanks! Could you add a test, for example in `DefaultTypeAdaptersTest`, that fails with the current code and passes with this change? You could use `assertThrows`: with the current code the test will fail because no exception will be thrown.

Please check again to ensure it meets the requirements.

**Review by eamonnmcmanus — APPROVED** [review]
Thanks! I am just checking this against Google's internal tests to see if they show up anything unexpected. If not, I'll go ahead and merge this.

**Comment by Marcono1234** [comment]
Should the new tests maybe be moved to `GsonBuilderTest` instead since they seem to only test `GsonBuilder` functionality?

Also, could you please change the title of this PR, there is an "Edit" button at the top right of the GitHub UI[^1]. The main change here seems to be about validating the date format, and issue 2472 might not be that closely related that it is worth mentioning it in the final commit message after merge. What do you think about something like this as title (based on what you wrote in the description)?
> Throw exception if `GsonBuilder.setDateFormat` input pattern is invalid

[^1]: Or alternatively, @eamonnmcmanus could you please adjust the message when merging?

File: gson/src/main/java/com/google/gson/GsonBuilder.java
Comment (Marcono1234) [inline gson/src/main/java/com/google/gson/GsonBuilder.java:599]: It appears `null` might actually be a valid `pattern` value to clear any custom pattern previously set for the builder. The changes here would not support that anymore.

@eamonnmcmanus, should `null` be supported?

Apparently no unit test is covering that yet, might be good to add one (or I can add one in a separate PR if desired).

Comment (Carpe-Wang) [reply gson/src/main/java/com/google/gson/GsonBuilder.java:599]: if pattern is null, here will  throw null point exception.

**Comment by Carpe-Wang** [comment]
> Should the new tests maybe be moved to `GsonBuilderTest` instead since they seem to only test `GsonBuilder` functionality?
> 
> Also, could you please change the title of this PR, there is an "Edit" button at the top right of the GitHub UI[1](#user-content-fn-1-eb8cfc75c4fc527652fd473304ede99c). The main change here seems to be about validating the date format, and issue 2472 might not be that closely related that it is worth mentioning it in the final commit message after merge. What do you think about something like this as title (based on what you wrote in the description)?
> 
> > Throw exception if `GsonBuilder.setDateFormat` input pattern is invalid
> 
> ## Footnotes
> 1. Or alternatively, @eamonnmcmanus could you please adjust the message when merging? [↩](#user-content-fnref-1-eb8cfc75c4fc527652fd473304ede99c)

I had change the title. Please check again to ensure it meets the requirements. In addition I'm not sure if I need to move the test cases to `GsonBuilderTest`, and I see a conflict with a small note that says "Only those with write access to this repository can merge pull requests." Do I need to resolve this conflict by myself?

**Comment by eamonnmcmanus** [comment]
>  I see a conflict with a small note that says "Only those with write access to this repository can merge pull requests." Do I need to resolve this conflict by myself?

Yes, you should resolve the conflict. I will merge the PR when it is ready.

I don't have a strong opinion on where the new tests should be. We may move them later. On the other hand, I agree with @Marcono1234  that we probably shouldn't change the behaviour for `setDateFormat(null)`. So the call to `new SimpleDateFormat` should be guarded by a `!= null` check.

File: gson/src/main/java/com/google/gson/internal/bind/util/ISO8601Utils.java
Comment (Carpe-Wang) [reply gson/src/main/java/com/google/gson/internal/bind/util/ISO8601Utils.java:172]: I am trying to solve that

**Pull request closed without merging**
